### ISSUE TITLE
pgwire: reading default_int_size can be racy

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#69451]
Setting `default_int_size` can potentially race with a pgwire race. Consider the sequence of steps:

* Run as a query `Query {"String": "SET default_int_size = 4"}`
* Send `Parse {"Query": "SELECT generate_series(1, 10)"}`

The second statement involves reading `default_int_size` off the `SessionData` from here: https://github.com/cockroachdb/cockroach/blob/16a8354aa86ff993846dda30f9b08729454fe1f8/pkg/sql/pgwire/conn.go#L434

However, the first statement still has to finish processing in `processCommandsAsync` for the latter to correct:
https://github.com/cockroachdb/cockroach/blob/16a8354aa86ff993846dda30f9b08729454fe1f8/pkg/sql/pgwire/conn.go#L298

However, if `SET default_int_size` is stalled and doesn't finish executing in `processCommandsAsync` before we read `PARSE` from the buffer, `PARSE` can use the old `default_int_size` value before it was set. However, we expect the `SET` to have already been completed before doing this. This causes the race.

# Further exacerbation

This is going to be further exacerbated by the implementation of `SET LOCAL`. Now, the race happens on the stack level because of the read of `UnqualifiedIntSize`. This happens when a begin query races against a parse:

```
==================
WARNING: DATA RACE
Write at 0x00c0048d4bb8 by goroutine 586:
  github.com/cockroachdb/cockroach/pkg/sql/sessiondata.(*Stack).Push()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sessiondata/session_data.go:234 +0x80f
  github.com/cockroachdb/cockroach/pkg/sql/sessiondata.(*Stack).PushTopClone()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sessiondata/session_data.go:239 +0x673
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInNoTxnState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1426 +0x624
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:111 +0x1405
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1686 +0x404
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1688 +0x1d14
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1610 +0x565
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:662 +0x134
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:647 +0x3f2

Previous read at 0x00c0048d4bb8 by goroutine 541:
  github.com/cockroachdb/cockroach/pkg/sql/sessiondata.(*Stack).Top()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sessiondata/session_data.go:226 +0x108
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).sessionData()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1515 +0xdf
  github.com/cockroachdb/cockroach/pkg/sql.ConnectionHandler.GetUnqualifiedIntSize()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:627 +0xbe
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:434 +0xd66
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:472 +0x984
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:160 +0x344
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:694 +0x1237
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2347 +0x16c
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:156 +0x11d

Goroutine 586 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:575 +0x1fd
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:298 +0x8b2
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:160 +0x344
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:694 +0x1237
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2347 +0x16c
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:156 +0x11d

Goroutine 541 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:153 +0x32c
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2343 +0x1a4
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x161
```

# Workaround

The workaround for this is to WaitForQuery when doing a `SET` of the `default_int_size`. This is an example of doing that: https://github.com/cockroachdb/cockroach/pull/69224/commits/57422100a7b23c3f64ab4c505e4ee1475f4aac19

# Repro

See #69448, run with `go test ./pkg/sql -run TestUnqualifiedIntSizeRace -test.v -race -count 30` 
```

func TestUnqualifiedIntSizeRace(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	ctx := context.Background()

	const (
		// select generates 10 rows of results which the test retrieves using ExecPortal
		selectStmt = "SELECT generate_series(1, 10)"

		// stmtBufMaxLen is the maximum length the statement buffer should be during
		// execution
		stmtBufMaxLen = 2

		// The name of the portal, used to get a handle on the statement buffer
		portalName = "C_1"
	)

	// This buffer is necessary for some reason.
	var stmtBuff *sql.StmtBuf
	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{
		Knobs: base.TestingKnobs{
			SQLExecutor: &sql.ExecutorTestingKnobs{
				// get a handle to the statement buffer during the Bind phase
				AfterExecCmd: func(_ context.Context, cmd sql.Command, buf *sql.StmtBuf) {
					switch tcmd := cmd.(type) {
					case sql.BindStmt:
						if tcmd.PortalName == "C_1" {
							stmtBuff = buf
						}
					default:
					}
				},
			},
		},
		Insecure: true,
	})
	defer s.Stopper().Stop(ctx)

	// Connect to the cluster via the PGWire client.
	p, err := pgtest.NewPGTest(ctx, s.SQLAddr(), security.RootUser)
	require.NoError(t, err)

	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 8"}`))
	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 4"}`))
	require.NoError(t, p.SendOneLine(`Parse {"Query": "SELECT generate_series(1, 10)"}`))
	require.NoError(t, p.SendOneLine(`Bind {"DestinationPortal": "C_1"}`))

	// wait for ready
	for i := 0; i < 2; i++ {
		until := pgtest.ParseMessages("ReadyForQuery")
		_, err = p.Until(false /* keepErrMsg */, until...)
		require.NoError(t, err)
	}

	_ = stmtBuff
}
```

Produces:
```
==================
WARNING: DATA RACE
Write at 0x00c00107adcc by goroutine 7:
  github.com/cockroachdb/cockroach/pkg/sql.(*sessionDataMutator).SetDefaultIntSize()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2611 +0x1f5
  github.com/cockroachdb/cockroach/pkg/sql.glob..func358()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/vars.go:350 +0x1cd
  github.com/cockroachdb/cockroach/pkg/sql.(*setVarNode).startExec.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_var.go:145 +0x221
  github.com/cockroachdb/cockroach/pkg/sql.(*sessionDataMutatorIterator).forEachMutatorError()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2549 +0x1c3
  github.com/cockroachdb/cockroach/pkg/sql.(*setVarNode).startExec()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/set_var.go:147 +0x41c
  github.com/cockroachdb/cockroach/pkg/sql.startExec.func2()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:509 +0xf5
  github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x99
  github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:289 +0x148
  github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0x104
  github.com/cockroachdb/cockroach/pkg/sql.walkPlan()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:43 +0x172
  github.com/cockroachdb/cockroach/pkg/sql.startExec()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:512 +0x187
  github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:142 +0x164
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:151 +0x19b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:86 +0x68
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedStatsCollectorImpl).Init()
      <autogenerated>:1 +0x57
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:234 +0x8f
  github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x72
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:233 +0x73
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:274 +0x10b
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:257 +0x3d3
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:535 +0x93d
  github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1276 +0x236
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1281 +0xa2f
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1017 +0xe25
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:708 +0x16f5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x12be
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1686 +0x404
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1688 +0x1d14
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1610 +0x565
  github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:663 +0x134
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:647 +0x3f2

Previous read at 0x00c00107adcc by goroutine 289:
  github.com/cockroachdb/cockroach/pkg/sql.ConnectionHandler.GetUnqualifiedIntSize()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:628 +0xd1
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:434 +0xd66
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:472 +0x984
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:160 +0x324
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:694 +0x11bc
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2347 +0x16c
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:156 +0x11d

Goroutine 7 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:575 +0x1fd
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:298 +0x8b2
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).serveConn()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:160 +0x324
  github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:694 +0x11bc
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1.1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2347 +0x16c
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:156 +0x11d

Goroutine 289 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:153 +0x32c
  github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).startServeSQL.func1()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:2343 +0x1a4
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      /Users/otan/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x161
```

# Future?

I think we should not allow `default_int_size` to be a session variable. It should be a cluster setting, with a fixed value on initialization of the pgwire connection.

**Comment by rafiss** [comment]
I think the repro above needs a bit of fixing. I'm able to repro the data race failure even when modifying that `TestUnqualifiedIntSizeRace` test like this:
```
-	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 8"}`))
-	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 4"}`))
+	require.NoError(t, p.SendOneLine(`Query {"String": "SET application_name = 'a'"}`))
+	require.NoError(t, p.SendOneLine(`Query {"String": "SET application_name = 'b'"}`))
```

Based on the output i get, it looks like the failure is because of the `_ = stmtBuff` read that's at the end of the test.

If I remove the stmtBuf entirely and change the test to:
```
func TestUnqualifiedIntSizeRace(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)

	ctx := context.Background()

	const (
		// select generates 10 rows of results which the test retrieves using ExecPortal
		selectStmt = "SELECT generate_series(1, 10)"

		// stmtBufMaxLen is the maximum length the statement buffer should be during
		// execution
		stmtBufMaxLen = 2

		// The name of the portal, used to get a handle on the statement buffer
		portalName = "C_1"
	)

	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{
		Knobs: base.TestingKnobs{
			SQLExecutor: &sql.ExecutorTestingKnobs{},
		},
		Insecure: true,
	})
	defer s.Stopper().Stop(ctx)

	// Connect to the cluster via the PGWire client.
	p, err := pgtest.NewPGTest(ctx, s.SQLAddr(), security.RootUser)
	require.NoError(t, err)

	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 8"}`))
	require.NoError(t, p.SendOneLine(`Query {"String": "SET default_int_size = 4"}`))
	require.NoError(t, p.SendOneLine(`Parse {"Query": "SELECT generate_series(1, 10)"}`))
	require.NoError(t, p.SendOneLine(`Bind {"DestinationPortal": "C_1"}`))

	// wait for ready
	for i := 0; i < 2; i++ {
		until := pgtest.ParseMessages("ReadyForQuery")
		_, err = p.Until(false /* keepErrMsg */, until...)
		require.NoError(t, err)
	}
}
```

then i do get the DefaultIntSize race with `make stressrace PKG=./pkg/sql TESTS=TestUnqualifiedIntSizeRace`.
```
WARNING: DATA RACE
Write at 0x00c0010084cc by goroutine 350:
  github.com/cockroachdb/cockroach/pkg/sql.(*sessionDataMutator).SetDefaultIntSize()
      /Users/rafiss/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec_util.go:2645 +0x1f5
```

I'm able to fix the race with this patch, which adds locks to SessionDataStack and using atomc.Load/Store for default int size.

(**probably for release-21.1 and before, we should backport the atomic.Load/Store change**)

```
diff --git a/pkg/sql/conn_executor.go b/pkg/sql/conn_executor.go
index fb0a615616..ec636af8bb 100644
--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -625,7 +625,8 @@ func (h ConnectionHandler) GetUnqualifiedIntSize() *types.T {
     if h.ex != nil {
         // The executor will be nil in certain testing situations where
         // no server is actually present.
-        size = h.ex.sessionData().DefaultIntSize
+        sd := h.ex.sessionData()
+        size = atomic.LoadInt32(&sd.DefaultIntSize)
     }
     return parser.NakedIntTypeFromDefaultIntSize(size)
 }

diff --git a/pkg/sql/exec_util.go b/pkg/sql/exec_util.go
index efa7737e76..49399b3fb5 100644
--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -2642,7 +2643,7 @@ func (m *sessionDataMutator) SetTemporarySchemaIDForDatabase(dbID uint32, tempSc
 }

 func (m *sessionDataMutator) SetDefaultIntSize(size int32) {
-    m.data.DefaultIntSize = size
+    atomic.StoreInt32(&m.data.DefaultIntSize, size)
 }

diff --git a/pkg/sql/sessiondata/session_data.go b/pkg/sql/sessiondata/session_data.go
index f59ac9d894..74f56055a8 100644
--- a/pkg/sql/sessiondata/session_data.go
+++ b/pkg/sql/sessiondata/session_data.go
@@ -201,6 +202,7 @@ func (s *SessionData) GetTemporarySchemaIDForDb(dbID uint32) (uint32, bool) {
 // affects the top of the stack.
 // There is always guaranteed to be one element in the stack.
 type Stack struct {
+    syncutil.RWMutex // and add s.Lock() / defer s.Unlock() to all the method.
     // Use an internal variable to prevent abstraction leakage.
     stack []*SessionData
 }
```

It's also fixed just by making `GetUnqualifiedIntSize()` stop reading sessiondata (as @otan suggested)
```
diff --git a/pkg/sql/conn_executor.go b/pkg/sql/conn_executor.go
index fb0a615616..ec636af8bb 100644
--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -625,7 +625,8 @@ func (h ConnectionHandler) GetUnqualifiedIntSize() *types.T {
     if h.ex != nil {
         // The executor will be nil in certain testing situations where
         // no server is actually present.
-        size = h.ex.sessionData().DefaultIntSize
     }
     return parser.NakedIntTypeFromDefaultIntSize(size)
 }
```

By the way, I noticed this comment in pgwire/conn.go which basically spells out the fact that there already is a race condition.
```
	// We need an intSizer, which we're ultimately going to get from the
	// authenticator once authentication succeeds (because it will actually be a
	// ConnectionHandler). Until then, we unfortunately still need some intSizer
	// because we technically might enqueue parsed statements in the statement
	// buffer even before authentication succeeds (because we need this go routine
	// to keep reading from the network connection while authentication is in
	// progress in order to react to the connection closing).
	var intSizer unqualifiedIntSizer = fixedIntSizer{size: types.Int}
```

So to summarize:
- for 21.1 and before, we should backport a change to make DefaultIntSize get accessed with `atomic.Load/Store`
- for 21.2 we either have to either:
  1. add a mutex around all SessionDataStack operations AND use atomic.Load/Store
  2. stop using a session setting for default_int_size. @otan could you explain a bit more why it's backward compatible to start ignoring the session setting?
  3. (haven't investigated yet) change pgwire/conn.go to stop using that `fixedIntSizer` and instead have it wait for the real `intSizer` to be sent over to it from the background goroutine. but seems like that would cause the pgwire `serveImpl` loop to block when we don't want it to.

**Comment by rafiss** [comment]
I'm marking this is a GA blocker for 21.2 since the SET LOCAL changes (SessionDataStack) makes it much easier to hit this.

**Comment by otan** [comment]
> stop using a session setting for default_int_size. @otan
<https://github.com/otan> could you explain a bit more why it's backward
compatible to start ignoring the session setting?

It's only backwards incompatible if we remove the cluster setting as well.
If we do this we still need it to be the cluster setting BUT we have to now
allow using SET default_int_size so it can't be modified during a session.

I have another fix in mind - make the session Data Stack only return the
bottom element for this read (and also for
https://github.com/cockroachdb/cockroach/issues/69508 )  to avoid this race
condition. I'll get to it on Monday.

**Comment by otan** [comment]
i don't think it'd be a GA blocker tbh, it *only* affects test and the consequences are similar to previous versions of CRDB.

**Comment by rafiss** [comment]
I think the scope of the bug is expanded -- now if the first command of a session is `BEGIN` you'll hit the bug too. (since BEGIN does the PushTopClone, which changes the `stack.Top()` pointer
